### PR TITLE
Upgrade samtools version.

### PR DESCRIPTION
**Pull request by DailyDreaming**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -17,28 +17,19 @@ tests:
...
+htslib:
+	wget https://github.com/samtools/htslib/releases/download/$(SAMTOOLS_VERSION)/htslib-$(SAMTOOLS_VERSION).tar.bz2 -O htslib.tar.bz2
 
-bcftools.tar.bz2:
-	git clone --depth 1 -b xbrianh-no-index https://github.com/xbrianh/bcftools
-	(cd bcftools ; rm -rf .git)
-	tar cjf bcftools.tar.bz2 bcftools
-	rm -rf bcftools
+bcftools:
+	wget https://github.com/samtools/bcftools/releases/download/$(SAMTOOLS_VERSION)/bcftools-$(SAMTOOLS_VERSION).tar.bz2 -O bcftools.tar.bz2
```
Comment (jessebrennan) [inline Makefile:29]: Changing the download method here is tangential to updating the version, right? In other words, it's a refactoring, that shouldn't change the functionality? If that's correct, it might make sense in a different PR, or at least in a separate commit with a message that makes that clear. If you don't feel like doing that, no worries, but I'd still like to know for my own sake lol.

File: setup.py [hunk 2]
```diff
@@ -20,14 +20,15 @@ def run(self):
         super().run()
         if not self.dry_run:
             try:
-                _run(["tar", "xjf", "htslib.tar.bz2", "-C", "build"])
-                _run(["tar", "xjf", "bcftools.tar.bz2", "-C", "build"])
-                _run(["tar", "xjf", "samtools.tar.bz2", "-C", "build"])
-                _run(["./configure"], cwd="build/htslib")
-                _run(["./configure"], cwd="build/samtools")
```
Comment (jessebrennan) [inline setup.py:27]: Why don't we need to run configure anymore?

**Review by jessebrennan — CHANGES_REQUESTED** [review]
I mostly have some questions, but I think I caught one minor bug.

A separate question, I don't understand why we need to include the tarballs in git. Shouldn't they just be downloaded when installing?

File: Makefile (hunk 1 above)
Comment (DailyDreaming) [reply Makefile:29]: This file is not the same as the repo and contains some generated dependencies.  They upload it separately from the auto-generated one that github produces.  See their release notes: https://github.com/samtools/bcftools/releases

File: setup.py [hunk 3]
```diff
@@ -20,14 +20,15 @@ def run(self):
...
+                os.makedirs('build/htslib', exist_ok=True)
+                _run(["tar", "-vxjf", "htslib.tar.bz2", "--directory=build/htslib", "--strip-components=1"])
+                os.makedirs('build/bcftools', exist_ok=True)
+                _run(["tar", "-vxjf", "bcftools.tar.bz2", "--directory=build/bcftools", "--strip-components=1"])
+                os.makedirs('build/samtools', exist_ok=True)
+                _run(["tar", "-vxjf", "samtools.tar.bz2", "--directory=build/samtools", "--strip-components=1"])
                 _run(["make"], cwd="build/htslib")
-                _run(["make"], cwd="build/bcftools")
                 _run(["make"], cwd="build/samtools")
+                _run(["make"], cwd="build/bcftools")
```
[the remark this answers is not in the document]
Comment (DailyDreaming) [reply setup.py:31]: This is needed to update the versions.

The `./configure` commands don't seem to be needed on the new versions, and were removed.

The internal folders have names like `htslib-1.15.1` instead of `htslib` and this prevents either 1. manually renaming the internal folder name and committing it each time we update or 2. renaming the directory once extracted.

All three libraries were updated at once, since they are all inter-related.

File: Makefile [hunk 4]
```diff
@@ -17,28 +17,19 @@ tests:
...
-package_samtools: clean_samtools samtools.tar.bz2 htslib.tar.bz2 bcftools.tar.bz2
+package_samtools: clean_samtools samtools htslib bcftools
 
-samtools.tar.bz2:
-	git clone --depth 1 -b $(SAMTOOLS_VERSION) https://github.com/samtools/samtools.git
-	(cd samtools ; rm -rf .git ; autoheader && autoconf)
-	tar cjf samtools.tar.bz2 samtools
-	rm -rf samtools
+samtools:
+	wget https://github.com/samtools/samtools/releases/download/$(SAMTOOLS_VERSION)/samtools-$(SAMTOOLS_VERSION).tar.bz2 -O samtools.tar.bz2
```
[the remark this answers is not in the document]
Comment (DailyDreaming) [reply Makefile:23]: Hmmm... I wasn't aware that make did caching, though I'm curious if caching is useful here?  Every time I manually run the command, I want to force update/download all files.  I don't see automating the command, or wanting caching even if I do.

What would be the benefit of caching here?

**Comment by DailyDreaming** [comment]
@jessebrennan I agree, the tarballs should be removed from git.  I was never a fan of that either, and had forgotten about it.

File: setup.py (hunk 2 above)
Comment (jessebrennan) [reply setup.py:27]: The docs recommend it https://github.com/samtools/samtools/blob/9472cc3d06c723ab91b38268421943afa813bbc4/INSTALL#L72 which seems like a good reason to me, but your argument is compelling so if you want to leave it out, fine.

Comment (DailyDreaming) [reply setup.py:27]: Eh, I thought about it and added it back in.

File: Makefile (hunk 4 above)
[the remark this answers is not in the document]
Comment (DailyDreaming) [reply Makefile]: These have now been removed altogether.

File: setup.py [hunk 5]
```diff
@@ -20,14 +20,18 @@ def run(self):
...
-                _run(["tar", "xjf", "htslib.tar.bz2", "-C", "build"])
-                _run(["tar", "xjf", "bcftools.tar.bz2", "-C", "build"])
-                _run(["tar", "xjf", "samtools.tar.bz2", "-C", "build"])
-                _run(["./configure"], cwd="build/htslib")
-                _run(["./configure"], cwd="build/samtools")
-                _run(["make"], cwd="build/htslib")
-                _run(["make"], cwd="build/bcftools")
-                _run(["make"], cwd="build/samtools")
+                for tool in ['samtools', 'htslib', 'bcftools']:
+                    if not os.path.exists(f'build/{tool}'):
```
Comment (jessebrennan) [inline setup.py:24]: I realize I've been flip-flopping here, but should we just download always? Like if we upgrade the version again (just by incrementing the number) the build won't change because the old version of the tool already exists.

Up to you here and I think practically it doesn't really matter, but wanted to mention it at least.

**Review by jessebrennan — APPROVED** [review]

File: setup.py (hunk 5 above)
Comment (DailyDreaming) [reply setup.py:24]: Good point.  I think I'll keep it this way for now, until I think of a better way.  Thanks for the review!
